src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -442,6 +459,84 @@ async def validate(self, data, schema_name, verrors, old=None):
                     f"The following security flavor(s) require NFSv4 to be enabled: {','.join(v4_sec)}."
                 )
 
+    @private
+    def validate_and_clean_share_hosts_intput(self, hosts, schema_name, verrors):
+        """
+        The host field can contain multiple 'host' entries that are space delimited
+        The validation is limited to detecting repeated 'host' names
+
+        This will return a list of hosts
+        """
+        new_host_list = []
+        for host_str in hosts:
+            # Trap space delimited hosts
+            if ' ' in host_str:
+                new_host_list.extend(shlex.split(host_str))
+                continue
+
+            # This should be a 'single' entry
+            new_host_list.append(host_str)
+
+        # Check for the 'everybody' (*) host entry.
+        if '*' in new_host_list:
+            if len(set(new_host_list)) > 1:
+                verrors.add(
+                    f"{schema_name}.hosts",
+                    "ERROR - '*', i.e. 'everybody', cannot be included with other entries"
+                )
+
+        # Check for duplicates
+        dups = list(set([x for x in new_host_list if new_host_list.count(x) >= 2]))
+        dup_hosts = ', '.join(dups)
+        if len(dups) > 0:
+            verrors.add(
+                f"{schema_name}.hosts",
+                f"ERROR - Duplicate host entries are not allowed: {dup_hosts}"
+            )
+
+        return new_host_list
+
+    @private
+    def validate_share_networks_input(self, networks, dns_cache, schema_name, verrors):
+        """
+        The network field is strictly limited to CIDR formats:
+        The input validator should enforce the CIDR format and a single address per entry.
+        This validation is limited to detecting repeats and overlapping subnets.
+        """
+        # Check for duplicates
+        if len(networks) != len(set(networks)):
+            dups = list(set([x for x in networks if networks.count(x) >= 2]))
+            verrors.add(
+                f"{schema_name}.networks",
+                f"ERROR - Duplicate network entries are not allowed: {dups}"
+            )
+        # Check for duplicates via hostname (which should be in dns_cache)
+        for IPaddr in networks:
+            IPinterface = ipaddress.ip_interface(IPaddr)
+
+            if str(IPinterface.ip) in list(dns_cache.values()):
```

Comment:
We're generating a list of the dns_cache values for each ip specified. We should generate this information outside the for loop just once.

```
dns_cache_values = list(dns_cache.values())
for IPaddr in networks:
    IPinterface = ipaddress.ip_interface(IPaddr)
    if str(IPinterface.ip) in dns_cache_values:
        ...